pyccel/parser/semantic.py

```diff
@@ -3785,212 +3911,263 @@ def _visit_FunctionDef(self, expr):
-        used_type_names = set(a.dtype for a in arg_annotations)
+        templatable_args = [unpack(a.annotation) for a in expr.arguments if isinstance(a.annotation, (SyntacticTypeAnnotation, UnionTypeAnnotation))]
+        arg_annotations  = [annot for a in templatable_args for annot in a if isinstance(annot, SyntacticTypeAnnotation)]
+        used_type_names  = set(a.dtype for a in arg_annotations)
         templates = {t: v for t,v in templates.items() if t in used_type_names}
 
+        tmp_templates = {}
+        new_expr_args = []
+        for a in expr.arguments:
+            if isinstance(a.annotation, UnionTypeAnnotation):
+                annotation = [aa.type_list if isinstance(aa, UnionTypeAnnotation) else [aa] for aa in a.annotation]
+                annotation = [aa for a in annotation for aa in a]
+            else:
+                annotation = [a.annotation]
+            if len(annotation)>1:
+                tmp_template_name = a.name + '_' + random_string(12)
+                tmp_templates[tmp_template_name] = UnionTypeAnnotation(*[self._visit(vi) for vi in annotation])
+                dtype_symb = PyccelSymbol(tmp_template_name, is_temp=True)
+                dtype_symb = SyntacticTypeAnnotation(dtype_symb)
+                var_clone = AnnotatedPyccelSymbol(a.var.name, annotation=dtype_symb)
+                new_expr_args.append(FunctionDefArgument(var_clone, bound_argument = a.bound_argument,
+                                        value=a.value, kwonly=a.is_kwonly, annotation=dtype_symb))
+            else:
+                new_expr_args.append(a)
+
+        templates.update(tmp_templates)
         template_combinations = list(product(*[v.type_list for v in templates.values()]))
         template_names = list(templates.keys())
         n_templates = len(template_combinations)
-
         # this for the case of a function without arguments => no headers
         interface_name = name
         interface_counter = 0
+        is_interface = n_templates > 1
+        annotated_args = [] # collect annotated arguments to check for argument incompatibility errors
+        for f in interface_funcs:
+            self.scope.remove_symbol(f.name)
 
         for tmpl_idx in range(n_templates):
-            # Change to syntactic FunctionDef scope to ensure get_expected_name is available
+
+            if is_interface:
+                name, interface_counter = self.scope.get_new_incremented_symbol(interface_name, interface_counter)
+
+            if function_call is not None and found_func:continue
+            if is_interface and function_call is not None and self.scope.find(name, 'functions'):
+                # the function has already been annotated
+                f = self.scope.find(name, 'functions')
+                assert f.is_annotated
+                funcs.append(f)
+                assert f is interface_funcs[0]
+                interface_funcs.pop(0)
```

Comment:
`funcs are the list of functions annotated in the current call of `_visit_FunctionDef` `interface_funcs` are the previously annotated functions of the interface, we should keep track of them and add them to `funcs` in an orderly way when we are in the loop

Reply:
Ok. Can we use names that make this distinction clearer? I suggest:
`funcs`->`new_semantic_funcs`
`interface_funcs`->`existing_semantic_funcs`

Unless `interface_funcs` also contains syntactic functions?

Reply:
no `interface_funcs` do not contain syntactic functions

Reply:
done in [d8e41de](https://github.com/pyccel/pyccel/pull/1728/commits/d8e41de46f64e73670c97a96cb32542e01e0858f)

Reply:
this chuck is removed in [3f07695](https://github.com/pyccel/pyccel/pull/1728/commits/3f07695f69a94a35bd3527899ec10cff00f61e10), because it was introduced to keep the order of the funcs, but the order is random and can change anyway, so we don't need to care about it